### ISSUE TITLE
Unable to clear title fields

### ISSUE THREAD

**Issue by nascentt** [desjardinsm/RicohAddressBook#3]
Due to input validation it's not possible to clear fields as "" and $null don't meet validation.

So I cannot do update-addressbokentry -id 1 Title1 OPQ -Title2 $null yo set title1 but clear title2

**Comment by desjardinsm** [comment]
The way that `Frequent`, `Title1`, `Title2`, and `Title3` work are a little bit funny, because of the way that those values are stored on the scanner. The values of those parameters on the scanner are actually saved as part of a single `tagId` property. That property is a comma-separated list of numbers: 1 indicates the "Add to Frequent" checkbox, 2 through 11 are for "Title 1" (the letters), 12 through 21 are for "Title 2" (that value minus 11), and 22 through 26 are for "Title 3" (that value minus 21).

This means that there's not really a good way in the `Update-AddressBookEntry` function to update the value of just one of those four without affecting the others, except to do a query as part of the update process (which could be quite expensive to do for each element in the pipeline). I ended up not implementing that operation for that reason, so providing one of those four parameters ends up resetting the ones you don't include. I tried to mention this in the `Get-Help` for `Update-AddressBookEntry`, but admittedly, it's not very obvious that you have to check that to find that out.

However, with the current implementation in mind, you should be able to just provide `-Title1 OPQ` and omit the `-Title2`, which should set "Title 1" to OPQ, and reset the "Add to Frequent", "Title 2", and "Title 3" values.

---
**Edit**: In addition, even if it was changed to perform a query against the scanner to find the current set of tags for the current pipeline entry, and then try to merge the changes in order to actually clear one of those fields, like you said, they have to allow a null or empty value.

The problem is that is that, at least for `Update-AddressBookEntry`, it has to know whether a value for that parameter was actually provided, in order to determine whether to leave that value as-is or modify it. For updating just a single entry, it's easy enough to determine whether a parameter was provided using `$PSBoundParameters.ContainsKey()`. However, when multiple objects are passed through the pipeline, including a value for a parameter in one object includes it in the `$PSBoundParameters` dictionary, even if it is not included in subsequent objects (its value will just be empty or null, depending on its type). This means that any naive checks will clear out the values of that property on any object that follows. The only real consistent way to tell whether it is included or not (that I have found) is to check against that default value. That means that it can't actually allow empty or null, which is why I included those separate `-Force*` switches, which actually clear the value if nothing is provided for them (by omitting them).

---

#### TL;DR

It's currently possible by providing `-Title1 OPQ` and omitting the `-Title2`, which should set "Title 1" to OPQ, and reset the "Add to Frequent", "Title 2", and "Title 3" values.

Changing it to be more intuitive is complicated.

**Comment by nascentt** [comment]
Fair enough.  So as I understand it currently doing an update-addressbookentry -title1 ABC will reset -title2 and frequent if it exists for that entry.

As long as there's a way to clear the value by API, then it's fine even if it is awkward.  I was just worried you had to use the web UI/printer tablet to clear them as there was no way via API.

**Comment by desjardinsm** [comment]
> So as I understand it currently doing an update-addressbookentry -title1 ABC will reset -title2 and frequent if it exists for that entry.

Yes, that is correct.

I'm contemplating an alternative where instead of passing Frequent/Title1/Title2/Title3 as separate parameters, they could be passed as a single object (something like `-Title @{ Frequent = $true; Title1 = 'AB' }`). That feels better to me, and it seems more obvious that they all go together and not including one in the object will reset that value in the address book.

This would be done as a second parameter set on the `Update-AddressBookEntry` function, so the existing parameters would be left alone.

The downside is that you won't get autocompletion when creating the object; it's still strongly typed and PowerShell knows how to coerce a hashtable to the underlying type if that's what you provide, but you don't get the autocompletion for it. I can make a type accelerator for the class and that might help a little with the property names (by allowing you to instantiate the class yourself, and then setting the properties manually; however, it does make it a bit more awkward to provide the object that way), but you don't get autocompletion for their values (just validation if you provide an invalid value).

I don't know how I feel about that.

**Comment by nascentt** [comment]
Yeah I'm not sure the change is necessary. Plus it might break backwards compatibility.

So which fields override others?  Cause frequent and title1 can both coexist.

**Comment by desjardinsm** [comment]
They can all technically coexist, it's just that if you provide one of the "Frequent"/"Title" fields (**Frequent**, **Title1**, **Title2**, or **Title3**), it sets the "tagId" property (which represents all four of them).

You can set all four if you want and they will all be set. However, as long as at least one of those four parameters are provided, the others that are not provided will be cleared, while those that are provided are set (as expected).

Also, not including any of them will not change any of them, so it is not possible to turn them all off, but that should be fine because the UI won't let you do that either.

**Comment by nascentt** [comment]
Appreciate the clarification.

Think I'm all good on my requests, everything I was trying to do is working out great thanks to your module.

Many thanks.
